node/src/components/linear_chain.rs

```diff
@@ -174,12 +168,37 @@ impl SignatureCache {
     }
 }
 
+#[derive(DataSize, Debug)]
+enum Signature {
+    Local(Box<FinalitySignature>),
+    External(Box<FinalitySignature>),
+}
+
+impl Signature {
+    fn to_inner(&self) -> &Box<FinalitySignature> {
+        match self {
+            Signature::Local(fs) => fs,
+            Signature::External(fs) => fs,
+        }
+    }
+
+    fn take(self) -> Box<FinalitySignature> {
+        match self {
+            Signature::Local(fs) | Signature::External(fs) => fs,
+        }
+    }
+
+    fn is_local(&self) -> bool {
+        matches!(self, Signature::Local(_))
+    }
+}
+
 #[derive(DataSize, Debug)]
 pub(crate) struct LinearChain<I> {
     /// The most recently added block.
     latest_block: Option<Block>,
     /// Finality signatures to be inserted in a block once it is available.
-    pending_finality_signatures: HashMap<PublicKey, HashMap<BlockHash, FinalitySignature>>,
+    pending_finality_signatures: HashMap<PublicKey, HashMap<BlockHash, Signature>>,
```

Comment:
Can our own ever end up there? We only sign a block if we already have that block… or can it happen that it's somehow not in storage yet when we handle our own signature?